libraries/GCS_MAVLink/GCS_FTP.cpp

```diff
@@ -0,0 +1,604 @@
+                        ((uint32_t *)reply.data)[0] = checksum;
+                        reply.opcode = FTP_OP::Ack;
+                        break;
+                    }
+                case FTP_OP::BurstReadFile:
+                    {
+                        // must actually be reading a file
+                        if (ftp.fd == -1) {
+                            ftp_error(reply, FTP_ERROR::FileNotFound);
+                            break;
+                        }
+
+                        // must have the file in read mode
+                        if ((ftp.mode != FTP_FILE_MODE::Read)) {
+                            ftp_error(reply, FTP_ERROR::Fail);
+                            break;
+                        }
+
+                        // seek to requested offset
+                        if (AP::FS().lseek(ftp.fd, request.offset, SEEK_SET) == -1) {
+                            ftp_error(reply, FTP_ERROR::FailErrno);
+                            break;
+                        }
+
+                        bool more_pending = true;
+                        const uint32_t transfer_size = 100;
+                        for (uint32_t i = 0; (i < transfer_size) && more_pending; i++) {
+                            // fill the buffer
+                            ssize_t read_bytes = AP::FS().read(ftp.fd, reply.data, sizeof(reply.data));
+                            if (read_bytes == -1) {
+                                ftp_error(reply, FTP_ERROR::FailErrno);
+                                more_pending = false;
+                                break;
+                            }
+
+                            if (read_bytes != sizeof(reply.data)) {
+                                // don't send any old data
+                                memset(reply.data + read_bytes, 0, sizeof(reply.data) - read_bytes);
+                            }
+
+                            if (read_bytes == 0) {
+                                ftp_error(reply, FTP_ERROR::EOF_);
+                                more_pending = false;
+                                break;
+                            }
+
+                            reply.opcode = FTP_OP::Ack;
+                            reply.offset = request.offset + i * sizeof(reply.data);
+                            reply.burst_complete = (i == (transfer_size - 1));
+                            reply.size = (uint8_t)read_bytes;
+
+                            while (!ftp.replies->push(reply)) { // we must fit the response, keep shoving it in
+                                hal.scheduler->delay(10);
+                            }
+
+                            // prep the reply to be used again
+                            reply.seq_number++;
+                        }
+
+                        break;
+                    }
+                case FTP_OP::TruncateFile:
+                case FTP_OP::Rename:
+                default:
+                    // this was bad data, just nack it
```

Comment:
This should be a panic in SITL.

Reply:
Why panic? The GCS asked for it, we can't do it. That would be like panicing on a `command_long` we didn't handle.